##WPŁYW UŻYWANYCH NARZĘDZI NA WARTOŚĆ ZAROBKÓW W DATA SCIENCE

Projekt ma za zadanie dokonać transformacji oraz agregacji danych dotyczących znajomości narzędzi Data Science wśród ankietowanych. Zostanie zbadany wpływ danych zmiennych na zarobki.

## Dane

Dane użyte do analizy pochodzą z platformy kaggle. Dotyczą one 2018 roku. Link: https://www.kaggle.com/kaggle/kaggle-survey-2018.

Osoby uczestniczące zostały poproszone w ankiecie o odpowiedzi na pytania co do znajomości pewnych narzędzi i technik Data Science.

Do analiz korzystamy z pliku multipleChoiceResponses, który zawiera w większości pytania i odpowiedzi wielokrotnego wyboru.

In [4]:
data = spark.read.format('com.databricks.spark.csv').options(inferSchema = 'true', header = 'true', delimiter = ',').load('/FileStore/tables/multipleChoiceResponses.csv')

Sprawdźmy najpierw z ilu wierszy i kolumn składają się dane.

In [6]:
print((data.count(), len(data.columns)))

(23860, 395)

Dane zawierają 23860 rekordów oraz aż 395 kolumn.

Zobaczmy teraz nazwy kolumn wraz z ich typami. Spróbujmy także odpowiedzieć na pytanie czemu kolumn jest aż tak dużo.

In [8]:
data.dtypes

Out[3]: [('Time from Start to Finish (seconds)', 'string'),
 ('Q1', 'string'),
 ('Q1_OTHER_TEXT', 'string'),
 ('Q2', 'string'),
 ('Q3', 'string'),
 ('Q4', 'string'),
 ('Q5', 'string'),
 ('Q6', 'string'),
 ('Q6_OTHER_TEXT', 'string'),
 ('Q7', 'string'),
 ('Q7_OTHER_TEXT', 'string'),
 ('Q8', 'string'),
 ('Q9', 'string'),
 ('Q10', 'string'),
 ('Q11_Part_1', 'string'),
 ('Q11_Part_2', 'string'),
 ('Q11_Part_3', 'string'),
 ('Q11_Part_4', 'string'),
 ('Q11_Part_5', 'string'),
 ('Q11_Part_6', 'string'),
 ('Q11_Part_7', 'string'),
 ('Q11_OTHER_TEXT', 'string'),
 ('Q12_MULTIPLE_CHOICE', 'string'),
 ('Q12_Part_1_TEXT', 'string'),
 ('Q12_Part_2_TEXT', 'string'),
 ('Q12_Part_3_TEXT', 'string'),
 ('Q12_Part_4_TEXT', 'string'),
 ('Q12_Part_5_TEXT', 'string'),
 ('Q12_OTHER_TEXT', 'string'),
 ('Q13_Part_1', 'string'),
 ('Q13_Part_2', 'string'),
 ('Q13_Part_3', 'string'),
 ('Q13_Part_4', 'string'),
 ('Q13_Part_5', 'string'),
 ('Q13_Part_6', 'string'),
 ('Q13_Part_7', 'string'),
 ('Q13_Part_8', 'string'),
 ('Q13_Part_9', 'string'),
 ('Q13_Part_10', 'string'),
 ('Q13_Part_11', 'string'),
 ('Q13_Part_12', 'string'),
 ('Q13_Part_13', 'string'),
 ('Q13_Part_14', 'string'),
 ('Q13_Part_15', 'string'),
 ('Q13_OTHER_TEXT', 'string'),
 ('Q14_Part_1', 'string'),
 ('Q14_Part_2', 'string'),
 ('Q14_Part_3', 'string'),
 ('Q14_Part_4', 'string'),
 ('Q14_Part_5', 'string'),
 ('Q14_Part_6', 'string'),
 ('Q14_Part_7', 'string'),
 ('Q14_Part_8', 'string'),
 ('Q14_Part_9', 'string'),
 ('Q14_Part_10', 'string'),
 ('Q14_Part_11', 'string'),
 ('Q14_OTHER_TEXT', 'string'),
 ('Q15_Part_1', 'string'),
 ('Q15_Part_2', 'string'),
 ('Q15_Part_3', 'string'),
 ('Q15_Part_4', 'string'),
 ('Q15_Part_5', 'string'),
 ('Q15_Part_6', 'string'),
 ('Q15_Part_7', 'string'),
 ('Q15_OTHER_TEXT', 'string'),
 ('Q16_Part_1', 'string'),
 ('Q16_Part_2', 'string'),
 ('Q16_Part_3', 'string'),
 ('Q16_Part_4', 'string'),
 ('Q16_Part_5', 'string'),
 ('Q16_Part_6', 'string'),
 ('Q16_Part_7', 'string'),
 ('Q16_Part_8', 'string'),
 ('Q16_Part_9', 'string'),
 ('Q16_Part_10', 'string'),
 ('Q16_Part_11', 'string'),
 ('Q16_Part_12', 'string'),
 ('Q16_Part_13', 'string'),
 ('Q16_Part_14', 'string'),
 ('Q16_Part_15', 'string'),
 ('Q16_Part_16', 'string'),
 ('Q16_Part_17', 'string'),
 ('Q16_Part_18', 'string'),
 ('Q16_OTHER_TEXT', 'string'),
 ('Q17', 'string'),
 ('Q17_OTHER_TEXT', 'string'),
 ('Q18', 'string'),
 ('Q18_OTHER_TEXT', 'string'),
 ('Q19_Part_1', 'string'),
 ('Q19_Part_2', 'string'),
 ('Q19_Part_3', 'string'),
 ('Q19_Part_4', 'string'),
 ('Q19_Part_5', 'string'),
 ('Q19_Part_6', 'string'),
 ('Q19_Part_7', 'string'),
 ('Q19_Part_8', 'string'),
 ('Q19_Part_9', 'string'),
 ('Q19_Part_10', 'string'),
 ('Q19_Part_11', 'string'),
 ('Q19_Part_12', 'string'),
 ('Q19_Part_13', 'string'),
 ('Q19_Part_14', 'string'),
 ('Q19_Part_15', 'string'),
 ('Q19_Part_16', 'string'),
 ('Q19_Part_17', 'string'),
 ('Q19_Part_18', 'string'),
 ('Q19_Part_19', 'string'),
 ('Q19_OTHER_TEXT', 'string'),
 ('Q20', 'string'),
 ('Q20_OTHER_TEXT', 'string'),
 ('Q21_Part_1', 'string'),
 ('Q21_Part_2', 'string'),
 ('Q21_Part_3', 'string'),
 ('Q21_Part_4', 'string'),
 ('Q21_Part_5', 'string'),
 ('Q21_Part_6', 'string'),
 ('Q21_Part_7', 'string'),
 ('Q21_Part_8', 'string'),
 ('Q21_Part_9', 'string'),
 ('Q21_Part_10', 'string'),
 ('Q21_Part_11', 'string'),
 ('Q21_Part_12', 'string'),
 ('Q21_Part_13', 'string'),
 ('Q21_OTHER_TEXT', 'string'),
 ('Q22', 'string'),
 ('Q22_OTHER_TEXT', 'string'),
 ('Q23', 'string'),
 ('Q24', 'string'),
 ('Q25', 'string'),
 ('Q26', 'string'),
 ('Q27_Part_1', 'string'),
 ('Q27_Part_2', 'string'),
 ('Q27_Part_3', 'string'),
 ('Q27_Part_4', 'string'),
 ('Q27_Part_5', 'string'),
 ('Q27_Part_6', 'string'),
 ('Q27_Part_7', 'string'),
 ('Q27_Part_8', 'string'),
 ('Q27_Part_9', 'string'),
 ('Q27_Part_10', 'string'),
 ('Q27_Part_11', 'string'),
 ('Q27_Part_12', 'string'),
 ('Q27_Part_13', 'string'),
 ('Q27_Part_14', 'string'),
 ('Q27_Part_15', 'string'),
 ('Q27_Part_16', 'string'),
 ('Q27_Part_17', 'string'),
 ('

Można rozszyfrować po nazwach kolumn, że są to odpowiedzi na zadane pytania. Niektóre pytania pojawiają się w więcej niż jednej kolumnie. Prawdopodobnie ankietowani mieli do wyboru kilka odpowiedzi na dane pytanie i odpowiedzi te zostały zapisane w różnych kolumnach.
 
Warto również zwrócić uwagę na to, że wszystkie kolumny są formatu tekstowego.

Spójrzmy teraz jak wyglądają dane, aby móc lepiej zrozumieć, jak są one zbudowane.

In [11]:
display(data.limit(10))

Time from Start to Finish (seconds) Q1 Q1_OTHER_TEXT Q2 Q3 Q4 Q5 Q6 Q6_OTHER_TEXT Q7 Q7_OTHER_TEXT Q8 Q9 Q10 Q11_Part_1 Q11_Part_2 Q11_Part_3 Q11_Part_4 Q11_Part_5 Q11_Part_6 Q11_Part_7 Q11_OTHER_TEXT Q12_MULTIPLE_CHOICE Q12_Part_1_TEXT Q12_Part_2_TEXT Q12_Part_3_TEXT Q12_Part_4_TEXT Q12_Part_5_TEXT Q12_OTHER_TEXT Q13_Part_1 Q13_Part_2 Q13_Part_3 Q13_Part_4 Q13_Part_5 Q13_Part_6 Q13_Part_7 Q13_Part_8 Q13_Part_9 Q13_Part_10 Q13_Part_11 Q13_Part_12 Q13_Part_13 Q13_Part_14 Q13_Part_15 Q13_OTHER_TEXT Q14_Part_1 Q14_Part_2 Q14_Part_3 Q14_Part_4 Q14_Part_5 Q14_Part_6 Q14_Part_7 Q14_Part_8 Q14_Part_9 Q14_Part_10 Q14_Part_11 Q14_OTHER_TEXT Q15_Part_1 Q15_Part_2 Q15_Part_3 Q15_Part_4 Q15_Part_5 Q15_Part_6 Q15_Part_7 Q15_OTHER_TEXT Q16_Part_1 Q16_Part_2 Q16_Part_3 Q16_Part_4 Q16_Part_5 Q16_Part_6 Q16_Part_7 Q16_Part_8 Q16_Part_9 Q16_Part_10 Q16_Part_11 Q16_Part_12 Q16_Part_13 Q16_Part_14 Q16_Part_15 Q16_Part_16 Q16_Part_17 Q16_Part_18 Q16_OTHER_TEXT Q17 Q17_OTHER_TEXT Q18 Q18_OTHER_TEXT Q19_Part_1 Q19_Part_2 Q19_Part_3 Q19_Part_4 Q19_Part_5 Q19_Part_6 Q19_Part_7 Q19_Part_8 Q19_Part_9 Q19_Part_10 Q19_Part_11 Q19_Part_12 Q19_Part_13 Q19_Part_14 Q19_Part_15 Q19_Part_16 Q19_Part_17 Q19_Part_18 Q19_Part_19 Q19_OTHER_TEXT Q20 Q20_OTHER_TEXT Q21_Part_1 Q21_Part_2 Q21_Part_3 Q21_Part_4 Q21_Part_5 Q21_Part_6 Q21_Part_7 Q21_Part_8 Q21_Part_9 Q21_Part_10 Q21_Part_11 Q21_Part_12 Q21_Part_13 Q21_OTHER_TEXT Q22 Q22_OTHER_TEXT Q23 Q24 Q25 Q26 Q27_Part_1 Q27_Part_2 Q27_Part_3 Q27_Part_4 Q27_Part_5 Q27_Part_6 Q27_Part_7 Q27_Part_8 Q27_Part_9 Q27_Part_10 Q27_Part_11 Q27_Part_12 Q27_Part_13 Q27_Part_14 Q27_Part_15 Q27_Part_16 Q27_Part_17 Q27_Part_18 Q27_Part_19 Q27_Part_20 Q27_OTHER_TEXT Q28_Part_1 Q28_Part_2 Q28_Part_3 Q28_Part_4 Q28_Part_5 Q28_Part_6 Q28_Part_7 Q28_Part_8 Q28_Part_9 Q28_Part_10 Q28_Part_11 Q28_Part_12 Q28_Part_13 Q28_Part_14 Q28_Part_15 Q28_Part_16 Q28_Part_17 Q28_Part_18 Q28_Part_19 Q28_Part_20 Q28_Part_21 Q28_Part_22 Q28_Part_23 Q28_Part_24 Q28_Part_25 Q28_Part_26 Q28_Part_27 Q28_Part_28 Q28_Part_29 Q28_Part_30 Q28_Part_31 Q28_Part_32 Q28_Part_33 Q28_Part_34 Q28_Part_35 Q28_Part_36 Q28_Part_37 Q28_Part_38 Q28_Part_39 Q28_Part_40 Q28_Part_41 Q28_Part_42 Q28_Part_43 Q28_OTHER_TEXT Q29_Part_1 Q29_Part_2 Q29_Part_3 Q29_Part_4 Q29_Part_5 Q29_Part_6 Q29_Part_7 Q29_Part_8 Q29_Part_9 Q29_Part_10 Q29_Part_11 Q29_Part_12 Q29_Part_13 Q29_Part_14 Q29_Part_15 Q29_Part_16 Q29_Part_17 Q29_Part_18 Q29_Part_19 Q29_Part_20 Q29_Part_21 Q29_Part_22 Q29_Part_23 Q29_Part_24 Q29_Part_25 Q29_Part_26 Q29_Part_27 Q29_Part_28 Q29_OTHER_TEXT Q30_Part_1 Q30_Part_2 Q30_Part_3 Q30_Part_4 Q30_Part_5 Q30_Part_6 Q30_Part_7 Q30_Part_8 Q30_Part_9 Q30_Part_10 Q30_Part_11 Q30_Part_12 Q30_Part_13 Q30_Part_14 Q30_Part_15 Q30_Part_16 Q30_Part_17 Q30_Part_18 Q30_Part_19 Q30_Part_20 Q30_Part_21 Q30_Part_22 Q30_Part_23 Q30_Part_24 Q30_Part_25 Q30_OTHER_TEXT Q31_Part_1 Q31_Part_2 Q31_Part_3 Q31_Part_4 Q31_Part_5 Q31_Part_6 Q31_Part_7 Q31_Part_8 Q31_Part_9 Q31_Part_10 Q31_Part_11 Q31_Part_12 Q31_OTHER_TEXT Q32 Q32_OTHER Q33_Part_1 Q33_Part_2 Q33_Part_3 Q33_Part_4 Q33_Part_5 Q33_Part_6 Q33_Part_7 Q33_Part_8 Q33_Part_9 Q33_Part_10 Q33_Part_11 Q33_OTHER_TEXT Q34_Part_1 Q34_Part_2 Q34_Part_3 Q34_Part_4 Q34_Part_5 Q34_Part_6 Q34_OTHER_TEXT Q35_Part_1 Q35_Part_2 Q35_Part_3 Q35_Part_4 Q35_Part_5 Q35_Part_6 Q35_OTHER_TEXT Q36_Part_1 Q36_Part_2 Q36_Part_3 Q36_Part_4 Q36_Part_5 Q36_Part_6 Q36_Part_7 Q36_Part_8 Q36_Part_9 Q36_Part_10 Q36_Part_11 Q36_Part_12 Q36_Part_13 Q36_OTHER_TEXT Q37 Q37_OTHER_TEXT Q38_Part_1 Q38_Part_2 Q38_Part_3 Q38_Part_4 Q38_Part_5 Q38_Part_6 Q38_Part_7 Q38_Part_8 Q38_Part_9 Q38_Part_10 Q38_Part_11 Q38_Part_12 Q38_Part_13 Q38_Part_14 Q38_Part_15 Q38_Part_16 Q38_Part_17 Q38_Part_18 Q38_Part_19 Q38_Part_20 Q38_Part_21 Q38_Part_22 Q38_OTHER_TEXT Q39_Part_1 Q39_Part_2 Q40 Q41_Part_1 Q41_Part_2 Q41_Part_3 Q42_Part_1 Q42_Part_2 Q42_Part_3 Q42_Part_4 Q42_Part_5 Q42_OTHER_TEXT Q43 Q44_Part_1 Q44_Part_2 Q44_Part_3 Q44_Part_4 Q44_Part_5 Q44_Part_6 Q45_Part_1 Q45_Part_2 Q45_Pa

Wygląda na to, że pierwszy wiersz danych zawiera w sobie wyjaśnienie nazw kolumn. Wyjaśnienie to jest zbyt długie, żeby nadawało się na nazwy kolumn, więc zastosowano podejście skrócenia nazw kolumn do litery Q, numeru pytania i ewentualnie jego rozróżnienia spośród różnych odpowiedzi na dane pytanie.

Warto również zwrócić uwagę, że niektóre pytania, na przykład Q11, składają się z kilku kolumn. Kolumny te oznaczają poszczególne odpowiedzi na dane pytanie. Odpowiedzi na poszczególne pytania nie ma specjalnie dużo, więc można założyć, że były to pytania zamknięte wielokrotnego wyboru. Co więcej, jeśli dana odpowiedź na pytanie jest twierdząca, w wierszu została umieszczona treść tej odpowiedzi (stąd wartości kolumn są tekstowe). Może się to przydać przy zmianie nazw kolumn na bardziej czytelne.

Można zauważyć jednak, że nie wszystkie kolumny dotyczą pytań zamkniętych, lecz część także otwartych. W analizie skupimy się jednak na zmiennych dotyczących pytań zamkniętych.

Zajmijmy się najpierw selekcją zmiennych. Wybierzmy jedynie najciekawsze pytania, według których porównywać będziemy następnie wyniki z zarobkami.

Wybrane pytania do dalszej analizy to:

* Q4 - edukacja
* Q7 - sektor, w którym dana osoba pracuje
* Q8 - doświadczenie w obecnej roli
* Q9 - zarobki
* Q11 - zajęcia w pracy
* Q15 - używane rozwiązania chmurowe
* Q16 - używane języki programowania
* Q17 - najczęściej używany język programowania
* Q19 - używane narzędzia machine learning
* Q20 - najczęściej używane narzędzie machine learning
* Q21 - używane narzędzia do wizualizacji
* Q22 - najczęściej używane narzędzie do wizualizacji
* Q29 - używane relacyjne bazy danych
* Q30 - używane narzędzia big data & analytics
* Q31 - typy danych

Zostawmy także zmienną Time from Start to Finish (seconds), która przyda się do wytypowania obserwacji, których nie można uznać za wiarygodne.

In [14]:
#tworzenie dataframe z nazwami kolumn
from pyspark.sql.types import StringType

columns_names = data.columns
columns_names = spark.createDataFrame(columns_names, StringType())

Stworzymy teraz 2 różne tabele. W jednej znajdą się pytania z dwiema cyframi, a w drugiej z jedną. Zabieg ten jest zrobiony, aby ułatwić wybór zmiennych (jeśli zmienna by się zaczynała od Q1, to zawierałaby także Q11, Q12, Q13,...). Następnie połączymy obie tabele.

In [16]:
#kolumny do zostawienia
element_list = ['Q11', 'Q15', 'Q16', 'Q17', 'Q19', 'Q20', 'Q21', 'Q22', 'Q29', 'Q30', 'Q31'] 

In [17]:
#funkcja do stosowania startswith dla wielu kolumn
from pyspark.sql.functions import col, lit
from functools import reduce

def start():
  starts_with = reduce(
    lambda x, y: x | y,
    [col("value").startswith(s) for s in element_list], 
    lit(False))
  return starts_with

In [18]:
#zmiana dataframe na listę
columns_to_retain = columns_names.filter(start())
columns_to_retain = columns_to_retain.rdd.flatMap(lambda x: x).collect()

In [19]:
data2 = data.select(columns_to_retain)

In [20]:
data1 = data.select('Time from Start to Finish (seconds)', 'Q2', 'Q4', 'Q7', 'Q8', 'Q9')

In [21]:
#dodanie takiego samego id do obu tabel, aby móc bez problemu je złączyć
from pyspark.sql.functions import monotonically_increasing_id

data1 = data1.withColumn("id", monotonically_increasing_id())
data2 = data2.withColumn("id", monotonically_increasing_id())

In [22]:
#tworzenie tymczasowych tabel w bazie Hive
data1.createOrReplaceTempView('data1')
data2.createOrReplaceTempView('data2')

In [23]:
#łączenie tabel
data = spark.sql("select * from data1 join data2 on data1.id = data2.id")

Odpowiednie kolumny do dalszych analiz zostały już wybrane.
Utwórzmy teraz słownik zawierający oznaczenie pytań oraz dane pytania, aby łatwo było nam się do nich odwoływać. Usuńmy także pierwszy wiersz z pełnymi nazwami danych. Zmieńmy również nazwę kolumny 'Time from Start to Finish (seconds)' na prostsze 'Time'

In [25]:
data_dict = data.limit(1)

In [26]:
data = data.withColumnRenamed('Time from Start to Finish (seconds)', 'Time')
data = data.filter(data['Time'] !='Duration (in seconds)')

Zajmijmy się teraz eliminacją zbędnych rekordów. Sprawdźmy jak wyglądają unikalne wartości dla zmiennej dotyczącej zarobków, od której będziemy później uzależniać kolejne zmienne.

In [28]:
data.select('Q9').distinct().show()

+--------------------+
 Q9|
+--------------------+
 400-500,000|
 40-50,000|
I do not wish to ...|
 null|
 80-90,000|
 300-400,000|
 200-250,000|
 125-150,000|
 150-200,000|
 250-300,000|
 100-125,000|
 30-40,000|
 500,000+|
 60-70,000|
 50-60,000|
 10-20,000|
 20-30,000|
 90-100,000|
 0-10,000|
 70-80,000|
+--------------------+

Jak widzimy w powyższej tabeli, pytanie 9 dotyczące zarobków zawiera w sobie 2 wartości niebędące przedziałem liczbowym, lecz odmową podania zarobków. Usuńmy więc takie wiersze, gdyż nie przedstawiają one zarobków ankietowanych.

In [30]:
data = data.filter(data['Q9'].isNotNull() & (data['Q9'] != 'I do not wish to disclose my approximate yearly compensation'))

In [31]:
data.select('Q9').distinct().show()
print('Liczba rekordów: ',data.count())

+-----------+
 Q9|
+-----------+
400-500,000|
 40-50,000|
 80-90,000|
300-400,000|
200-250,000|
125-150,000|
150-200,000|
250-300,000|
100-125,000|
 30-40,000|
 500,000+|
 60-70,000|
 50-60,000|
 10-20,000|
 20-30,000|
 90-100,000|
 0-10,000|
 70-80,000|
+-----------+

Liczba rekordów: 15429

Udało nam się pozbyć wierszy niezawierających przedziałów zaroboków ankietowanych. Co więcej z ponad 23 tysięcy wierszy zostało teraz jedynie ponad 15 tysięcy. 

W dalszej części projektu będziemy liczyli zarobki danej osoby jako średnią z przedziału aktualnych wartości kolumny Q9. Najwyższa wartość zarobków (500,000+) jest jednak przedziałem otwartym, więc nie nadaje się ona do naszej koncepcji. Zbadajmy więc głębiej informacje o takich rekordach, zaczynając od liczby takich obserwacji.

In [33]:
print('Liczba rekordów z wartością rocznych zarobków wynoszącą ponad 500 tysięcy dolarów:', data.select('Q9').filter(data.Q9 == '500,000+').count())

print('Procent liczby wszystkich rekordów:', round(data.select('Q9').filter(data.Q9 == '500,000+').count() * 100 / (data.count()), 2), '%')

Liczba rekordów z wartością rocznych zarobków wynoszącą ponad 500 tysięcy dolarów: 63
Procent liczby wszystkich rekordów: 0.41 %

63 wartości stanowi zaledwie 0.4% wszystkich rekordów zbioru. Taka część zbioru nie powinna zmienić w dużym stopniu wyników, więc możemy usunąć wiersze z taką wartością.

Dodatkowo taka wartość zarobków nie jest dobrym wskaźnikiem w ustaleniu, które narzędzia determinują zarobki w Data Science, gdyż zarobki najlepiej zarabiających zależą w dużym stopniu także od innych czynników (takie zarobki są raczej zarezerwowane dla szczebla kierowniczego) oprócz znajomości konkretnych narzędzi Data Science.

In [35]:
data = data.filter(data.Q9 != '500,000+')

Sprawdźmy teraz jak przedstawiają się podstawowe statystyki czasu wypełnienia ankiety. Żeby to osiągnąć, zmieńmy jednak najpierw rodzaj kolumny na numeryczny

In [37]:
from pyspark.sql.types import DoubleType
data = data.withColumn("Time", data.Time.cast(DoubleType()))

In [38]:
data.describe(['Time']).show()

+-------+------------------+
summary| Time|
+-------+------------------+
 count| 15366|
 mean|13046.193283873487|
 stddev| 48663.05510151093|
 min| 54.0|
 max| 595857.0|
+-------+------------------+

Jak widzimy, długość trwania wypełnienia ankiety dość się różniła pomiędzy poszczególnymi osobami. Pewna osoba wypełniła ankietę w ciągu 54 sekund, a najdłużej ankieta była wypełniana prawie 7 dni. Ankieta mogła być długo wypełniana, gdy ktoś zostawił włączone okno z ankietą na kilka dni, lecz czas wypełnienia poniżej minuty zaczyna budzić już podejrzenia. Wśród tak rozbieżnych długości trwania średnia nie jest dobrą metodą pokazującą średni czas wypełnienia ankiety (ponad 3 godziny). Posłużmy się więc kwartylami, które nie są aż tak podatne na wartości odstające.

In [40]:
print('Kwartyl 25:' , round(data.approxQuantile("Time", [0.25], 0)[0]/60, 2), 'minut\nMediana:', round(data.approxQuantile("Time", [0.5], 0)[0]/60,2), 'minut\nKwartyl 75:', round(data.approxQuantile("Time", [0.75], 0)[0]/60,2), 'minut')

Kwartyl 25: 12.23 minut
Mediana: 18.73 minut
Kwartyl 75: 34.12 minut

Jak możemy zauważyć, mediana wypełnienia ankiety wyniosła około 19 minut. Prawie 25% wartości było niższe niż 12 minut, a 25% wartości było większe niż 34 minuty. Ankieta wypełniana przez 25% najszybszych osób w ciągu maksymalnie 12 minut oznacza, że nie mogła być ona wypełniona rzetelnie przez osoby wypełniające ją w ciągu 2 minut. Usuńmy więc obserwacje z takimi wartościami, a następnie usuńmy też samą kolumnę czasu wypełnienia ankiety, gdyż jest już zbędna.

Usuńmy także również zmienne dotyczące id obserwacji, gdyż nie są nam one już do niczego potrzebne.

In [42]:
print('Liczba ankiet wypełnionych w ciągu dwóch minut:', data.filter(data.Time <= 120).count())
data = data.filter(data.Time > 120)
data = data.drop('Time', 'id')
print('Nowa liczba rekordów:', data.count())

Liczba ankiet wypełnionych w ciągu dwóch minut: 278
Nowa liczba rekordów: 15088

Pozbyliśmy się mało wiarygodnych wypełnień ankiety ze względu na czas ich wypełnienia.

Usuńmy teraz jeszcze zmienne oznaczające inną odpowiedź na pytanie, niż założone. Odpowiedzi na pytania jest wystarczająco dużo, żeby dokonać analizy danych najważniejszych narzędzi używanych w Data Science.

In [44]:
#Usuńmy kolumny kończące się na other_text, bo przeszkadzają
data = data.drop('Q11_OTHER_TEXT', 'Q15_OTHER_TEXT', 'Q16_OTHER_TEXT', 'Q17_OTHER_TEXT', 'Q19_OTHER_TEXT', 'Q20_OTHER_TEXT', 'Q21_OTHER_TEXT',  'Q26_OTHER_TEXT', 'Q22_OTHER_TEXT', 'Q29_OTHER_TEXT', 'Q30_OTHER_TEXT', 'Q31_OTHER_TEXT', 'Q34_OTHER_TEXT', 'Q41_OTHER_TEXT', 'Q47_OTHER_TEXT')

####Zmiana nazw kolumn

Nazwijmy kolumny, których rekordy zawierają tylko odpowiedzi na pytania oraz wartości null, za pomocą wypełnianych przez nie wartości. Postąpimy w ten sposób, że wśród kolumn sprawdzimy wartości pierwszych 1000 wierszy, a następnie wybierzemy wartości unikatowe, spośród których wybierzemy najczęściej powtarzającą się wartość niebędącą nullem. Taki sposób postępowania gwarantuje nam:

* zapytanie przelicza się dla 1000 wierszy zamiast ponad 15 tysięcy (wszystkich wierszy) dla każdej kolumny, dzięki czemu wykonuje się szybciej
* 1000 rekordów powinno być wystarczającą ilością, żeby w każdej z kolumn znalazły się wartości dla danej zmiennej

In [46]:
#lista pytań z wieloma odpowiedziami zamkniętymi
element_list =  ['Q11', 'Q15', 'Q16', 'Q19', 'Q21', 'Q26', 'Q29', 'Q30', 'Q31']

In [47]:
columns_list = data.columns
columns_list = spark.createDataFrame(columns_list, StringType())

columns_to_rename = columns_list.filter(start())
columns_to_rename = columns_to_rename.rdd.flatMap(lambda x: x).collect()

In [48]:
#Stwórzmy jeszcze na wszelki wypadek kopię danych przed zmianą nazw kolumn
data_copy = data

In [49]:
for i in range(len(columns_to_rename)):
  val_from_col = data.limit(1000).select(columns_to_rename[i]).where(data[columns_to_rename[i]].isNotNull()).distinct()
  val_to_list = val_from_col.rdd.flatMap(lambda x: x).collect()
  data = data.withColumnRenamed(columns_to_rename[i], val_to_list[0])

Utworzone kolumny wyglądają w większości poprawnie, jednak można zauważyć kilka błędów:

* wiele kolumn o nazwie Other
* wiele kolumn o nazwie None

Kolumn nazwanych Others nie będziemy brać pod uwagę, gdyź możliwe odpowiedzi na pytania są tak szczegółowe, że narzędzia, które nie pasowały do żadnej z grup powinny być zupełnie niszowe.

Nie będziemy brać również pod uwagę kolumn None, gdyż zależy nam na wiedzy o wpływie danych narzędzi na zarobki, a nie ich braku.

Nie będziemy jednak usuwać tych kolumn, aby słownik tabel miał taką samą strukturę, jak nowo nazwane kolumny.

Zajmijmy się teraz jeszcze nazwami kolumn przyjmującymi wiele możliwych wartości.

In [52]:
data = data.withColumnRenamed('Q2', 'Age')
data = data.withColumnRenamed('Q4', 'Degree')
data = data.withColumnRenamed('Q7', 'Sector')
data = data.withColumnRenamed('Q8', 'Experience')
data = data.withColumnRenamed('Q9', 'Salary')
data = data.withColumnRenamed('Q17', 'Most often used programming language')
data = data.withColumnRenamed('Q22', 'Most often used visualization framework')
data = data.withColumnRenamed('Q20', 'Most often used maching learning framework')

Wszystkie kolumny mają już odpowiednie nazwy. Możemy się teraz zająć przygotowaniami do agregacji danych.

####Agregacja danych

Stwórzmy teraz ze zmiennej Salary zmienne pokazujące minimalną, maksymalną oraz średnią wartość zarobków danej osoby. Do porównań z innymi zmiennymi użyjemy wartości średniej, gdyż prawdopodobnie z najlepszym przybliżeniem pokazuje ona faktyczny stan zarobków ankietowanych.

In [56]:
from pyspark.sql.functions import split
from pyspark.sql.functions import mean

split_col = split(data['Salary'], '-')
data = data.withColumn('Salary_min', split_col.getItem(0))
data = data.withColumn('Salary_max', split_col.getItem(1))

In [57]:
display(data.select('Salary', 'Salary_min', 'Salary_max').limit(10))

Salary Salary_min Salary_max 10-20,000 10 20,000 0-10,000 0 10,000 0-10,000 0 10,000 10-20,000 10 20,000 0-10,000 0 10,000 20-30,000 20 30,000 125-150,000 125 150,000 30-40,000 30 40,000 30-40,000 30 40,000 10-20,000 10 20,000

Tworzenie nowych zmiennych z minimalnymi i maksymalnymi wartościami zarobków przebiegło pomyślnie. Kolumna Salary_min powinna być jednak przemnożona przez tysiąc, a kolumna Salary_max z kolei powinna pozbyć się znaku przecinka z wartości. Zmieńmy więc te wartości, a następnie zamieńmy jeszcze typy nowych kolumn na liczbowy. Następnie będziemy mogli stworzyć zmienną dotyczącą średniej wartości zarobków.

In [59]:
from pyspark.sql.functions import regexp_replace
data = data.withColumn('Salary_max', regexp_replace('Salary_max', ',', ''))
data = data.withColumn('Salary_min', data.Salary_min * 1000)

In [60]:
from pyspark.sql.types import IntegerType
data = data.withColumn("Salary_min", data.Salary_min.cast(IntegerType()))
data = data.withColumn("Salary_max", data.Salary_max.cast(IntegerType()))
data = data.withColumn('Salary_avg', (data.Salary_min + data.Salary_max) / 2)

In [61]:
display(data.select('Salary', 'Salary_min', 'Salary_max', 'Salary_avg').limit(10))

Salary Salary_min Salary_max Salary_avg 10-20,000 10000 20000 15000.0 0-10,000 0 10000 5000.0 0-10,000 0 10000 5000.0 10-20,000 10000 20000 15000.0 0-10,000 0 10000 5000.0 20-30,000 20000 30000 25000.0 125-150,000 125000 150000 137500.0 30-40,000 30000 40000 35000.0 30-40,000 30000 40000 35000.0 10-20,000 10000 20000 15000.0

Kolumny Salary_min, Salary_max oraz Salary_avg zostały utworzone.

Możemy więc w końcu przejść do agregacji i wizualizacji danych.

Utwórzmy najpierw funkcje tworzące podsumowanie liczby unikatowych wartości dla danej kolumny oraz średnią wartość zarobków dla każdej z wartości danej cechy.

In [64]:
from pyspark.sql.functions import bround
from pyspark.sql.functions import desc

#funkcje dla zmiennych z pojedyncza kolumną
def val_counts(var):
  display(data.select(data.columns[var]).groupBy(data.columns[var]).count().sort(desc('count')))
def avg_salary(var):
  query = data.select(data.columns[var], 'Salary_avg').groupBy(data.columns[var]).avg('Salary_avg')
  display(query.select(data.columns[var], bround('avg(Salary_avg)').alias('Average salary')).sort((data.columns[var])))
  
#funkcje dla zmiennych znajdujących się w wielu kolumnach  
chart_names = ['Activities at work', 'Cloud providers', 'Programming languages', 
'Machine learning frameworks', 'Visualization tools', 'Relational databases', 
'Big Data and analytics', 'Data types']

def val_cols_counts(first_col, last_col):
  results = []
  for question in range(first_col, last_col):
    results.append([data.columns[question], data.select(data.columns[question]).filter(data[data.columns[question]] == data.columns[question]).count()])
    display(spark.createDataFrame(results).toDF(chart_names[chart], 'count').sort(desc('count')))    
  
def avg_cols_salary(first_col, last_col):
  results = []
  for question in range(first_col, last_col):
    query = data.select(data.columns[question], 'Salary_avg').filter(data[data.columns[question]] == data.columns[question]).groupBy(data.columns[question]).avg('Salary_avg')
    results.append([data.columns[question], query.select(bround('avg(Salary_avg)').alias('Average salary')).sort((data.columns[question])).collect()[0]['Average salary']])
    display(spark.createDataFrame(results).toDF(chart_names[chart], 'Average salary').sort(chart_names[chart]))    

###Wiek

In [66]:
val_counts(0)

Age count 25-29 4199 22-24 2812 30-34 2777 35-39 1610 18-21 1291 40-44 981 45-49 594 50-54 396 55-59 217 60-69 177 70-79 25 80+ 9

In [67]:
avg_salary(0)

Age Average salary 18-21 11069.0 22-24 22676.0 25-29 39499.0 30-34 59672.0 35-39 70438.0 40-44 76748.0 45-49 89844.0 50-54 97298.0 55-59 112177.0 60-69 112938.0 70-79 85900.0 80+ 96667.0

Jak można się było spodziewać, wraz z wiekiem (i doświadczeniem pracownika) rosną również jego zarobki, jednak zaczynają one już spadać dla osób od 70 roku życia. Badanych osób 70+ było jednak bardzo mało, więc nie można uznać wyników dla osób powyżej 70 roku życia za istotne. Logicznie rzecz biorąc jest niewiele osób zajmujących się Data Science w takim wieku, więc trudno na tej podstawie generalizować wyniki.

Najwięcej ankietowanych było w przedziale od 18 do 44 lat, więc również z tego powodu zarobki dla osób w tym przedziale wiekowym będą się kształtować bliżej średniej.

Generalny trend wzrosotowy wzraz ze wzrostem wynagrodzeń wydaje się być zdecydowanie widoczny.

###Tytuł naukowy

In [70]:
val_counts(1)

Degree count Master’s degree 7146 Bachelor’s degree 4271 Doctoral degree 2471 Some college/university study without earning a bachelor’s degree 572 Professional degree 381 No formal education past high school 124 I prefer not to answer 123

In [71]:
avg_salary(1)

Degree Average salary Bachelor’s degree 37955.0 Doctoral degree 65638.0 I prefer not to answer 33598.0 Master’s degree 51199.0 No formal education past high school 45726.0 Professional degree 41115.0 Some college/university study without earning a bachelor’s degree 38330.0

Jak można zauważyć z wykresu, najwięcej zarabiają osoby z tytułem Doktora (średnio około 15 tysięcy dolarów rocznie więcej niż w przypadku tytułu magistra). Stopień licencjata oznacza kolejny spadek wynagrodzenia o 14 tysięcy dolarów.

###Sektor

In [74]:
val_counts(2)

Sector count Computers/Technology 4113 I am a student 2144 Academics/Education 1923 Accounting/Finance 1155 Other 742 Online Service/Internet-based Services 706 Medical/Pharmaceutical 577 Government/Public Service 516 Insurance/Risk Assessment 450 Manufacturing/Fabrication 429 Marketing/CRM 378 Retail/Sales 351 Energy/Mining 338 Broadcasting/Communications 316 Online Business/Internet-based Sales 268 Shipping/Transportation 235 Non-profit/Service 152 Hospitality/Entertainment/Sports 149 Military/Security/Defense 146

In [75]:
avg_salary(2)

Sector Average salary Academics/Education 34887.0 Accounting/Finance 63965.0 Broadcasting/Communications 56108.0 Computers/Technology 53628.0 Energy/Mining 62478.0 Government/Public Service 60329.0 Hospitality/Entertainment/Sports 73456.0 I am a student 10976.0 Insurance/Risk Assessment 69922.0 Manufacturing/Fabrication 57570.0 Marketing/CRM 57354.0 Medical/Pharmaceutical 69844.0 Military/Security/Defense 66747.0 Non-profit/Service 46217.0 Online Business/Internet-based Sales 50868.0 Online Service/Internet-based Services 64377.0 Other 58908.0 Retail/Sales 53754.0 Shipping/Transportation 63819.0

Jak widzimy, największe zarobki zdają się osiągać Data Scientists z branży rozrywkowej oraz sportu. Stanowią oni jednak niewielką część wszystkich badanych. Kolejne miejsca po nich zajmują branże ubezpieczeniowa, medyczna oraz finansowa.

Zdecydowanie najmniej natomiast zarabiają studenci, a następnie nauczyciele akademiccy, pracownicy organizacji non-profit oraz e-biznes.

###Doświadczenie

In [78]:
val_counts(3)

Experience count 0-1 3767 1-2 2762 5-10 1938 2-3 1934 3-4 1284 4-5 1125 10-15 1104 15-20 582 20-25 267 30 + 131 25-30 119 null 75

In [79]:
avg_salary(3)

Experience Average salary null 10033.0 0-1 32200.0 1-2 39914.0 10-15 72237.0 15-20 72921.0 2-3 45243.0 20-25 99448.0 25-30 108929.0 3-4 46986.0 30 + 118702.0 4-5 54462.0 5-10 61360.0

Jak się można było spodziewać, im większe doświadczenie, tym wyższe zarobki. Trend ten wydaje się stale utrzymywać.

###Czynności wykonywane w pracy

In [82]:
chart = 0
val_cols_counts(5, 11)

Activities at work count Analyze and understand data to influence product or business decisions 7411 Build prototypes to explore applying machine learning to new areas 5683 Build and/or run a machine learning service that operationally improves my product or workflows 4209 Build and/or run the data infrastructure that my business uses for storing, analyzing, and operationalizing data 4120 Do research that advances the state of the art of machine learning 3551 None of these activities are an important part of my role at work 2494

In [83]:
avg_cols_salary(5, 11)

Activities at work Average salary Analyze and understand data to influence product or business decisions 59945.0 Build and/or run a machine learning service that operationally improves my product or workflows 62860.0 Build and/or run the data infrastructure that my business uses for storing, analyzing, and operationalizing data 62749.0 Build prototypes to explore applying machine learning to new areas 64415.0 Do research that advances the state of the art of machine learning 50754.0 None of these activities are an important part of my role at work 34747.0

Wyniki pokazują, że ludzie, którzy zajmują się w pracy takimi zadaniami jak analizą danych pod kątem biznesowym, budową usług machine learning, tworzeniem infrastruktury pod gromadzenie i analizę danych oraz budową prototypów zarabiają średnio podobną kwotę pieniędzy. Mniej opłacany jest research (około 50 tysięcy dolarów), a osoby, które nie uważają powyższych czynności za ważne w ich w pracy zarabiają kolejne 15 tysięcy mniej.

### Usługi chmurowe

In [86]:
chart = chart + 1
val_cols_counts(12, 18)

Cloud providers count Amazon Web Services (AWS) 5830 I have not used any cloud providers 4744 Google Cloud Platform (GCP) 3581 Microsoft Azure 2917 IBM Cloud 868 Alibaba Cloud 356

In [87]:
avg_cols_salary(12, 18)

Cloud providers Average salary Alibaba Cloud 31910.0 Amazon Web Services (AWS) 65327.0 Google Cloud Platform (GCP) 56525.0 I have not used any cloud providers 36942.0 IBM Cloud 55726.0 Microsoft Azure 61166.0

Z wyników ankiet wynika, że najlepiej opłacani są pracownicy koszystający z usług chmurowych AWS. Zaraz po nich są Azure, IBM Cloud oraz Google Cloud. Korzystanie z Alibaba Cloud jest mniej cenione (prawie 2 razy mniejsze wynagrodzenie niż za korzystanie z AWS), nawet mniejsze od osób niewykorzystujących usług chmurowych. Prawdopodobnie wynika to z używania Alibaby przez firmy azjatyckie, w których wynagrodzenie pracowników jest dużo mniejsze.

### Używane języki programowania

In [90]:
#Zmieniona nazwa i wartości kolumny, ponieważ znaki występujące w zmiennej wywoływały błąd
from pyspark.sql.functions import when

data = data.withColumnRenamed(data.columns[31], 'CSharpNET')
data = data.withColumn('CSharpNET', when(data.CSharpNET == 'C#/.NET', 
lit('CSharpNET')))

In [91]:
chart = chart + 1
val_cols_counts(19, 35)

Programming languages count Python 11316 SQL 6196 R 4965 C/C++ 2813 Java 2680 Javascript/Typescript 2372 Bash 2129 MATLAB 1748 CSharpNET 1187 Visual Basic/VBA 956 PHP 832 SAS/STATA 798 Scala 635 Go 281 Ruby 246 Julia 165

In [92]:
avg_cols_salary(19, 35)

Programming languages Average salary Bash 67265.0 C/C++ 38997.0 CSharpNET 49844.0 Go 71504.0 Java 44972.0 Javascript/Typescript 52531.0 Julia 64818.0 MATLAB 37398.0 PHP 37930.0 Python 50202.0 R 56309.0 Ruby 68994.0 SAS/STATA 61109.0 SQL 57580.0 Scala 72024.0 Visual Basic/VBA 54417.0

Jak widzimy na powyższym wykresie, najlepiej opłacanymi językami programowania w Data Science według ankiety okazują się być Scala, Go, Ruby i Bash.

Z kolei C/C++, MATLAB, PHP, a nawet Java nie wpływają w dobry sposób na zarobki respondentów.

Na uwagę zasługują także wyższe płace dla znajomości R niż dla Pythona. Może to wynikać z tego, że język Python staje się coraz bardziej powszechnym językiem w Data Science jako pierwszy wybór, więc zna go więcej osób (również początkujący wydają się na nim skupiać). Jak możemy zobaczyć w liczbie wypełnień ankiety, wypełnień dla Pythona jest ponad 2 razy więcej niż R.

### Najczęściej używany język programowania

In [95]:
val_counts(37)

Most often used programming language count Python 5850 null 4033 R 1520 SQL 996 Java 620 C/C++ 462 C#/.NET 329 Javascript/Typescript 308 MATLAB 246 SAS/STATA 178 PHP 149 Visual Basic/VBA 95 Other 92 Scala 81 Bash 45 Ruby 42 Go 35 Julia 7

In [96]:
avg_salary(37)

Most often used programming language Average salary null 40692.0 Bash 63333.0 C#/.NET 52204.0 C/C++ 39573.0 Go 75071.0 Java 47524.0 Javascript/Typescript 45235.0 Julia 52143.0 MATLAB 37764.0 Other 62799.0 PHP 35201.0 Python 49906.0 R 57910.0 Ruby 65714.0 SAS/STATA 66671.0 SQL 62525.0 Scala 78395.0 Visual Basic/VBA 54947.0

Jak wynika z wykresu, zdecydowanie najwyższe zarobki występują dla osób, których głównym językiem programowanie jest Scala oraz Go. Ruby i Bash oznaczają już spadek zarobków o około 20 tysięcy dolarów. Otrzymaliśmy więc doprecyzowanie do wykresów z ogólnie wykorzystywanymi językami programowania, w których Scala, Go, Ruby i Bash były na podobnym poziomie zarobków.

Wciąż najmniej opłacanymi językami w Data Science wydają siębyć C/C++, MATLAB oraz PHP. Wyniki dla Javy jako głównego języka programowania zaczynają się już prezentować trochę lepiej.

###Narzędzia machine learning

In [99]:
chart = chart + 1
val_cols_counts(38, 55)

Machine learning frameworks count Scikit-Learn 9049 TensorFlow 7327 Keras 6040 randomForest 4463 Xgboost 4286 PyTorch 2810 Caret 1757 Spark MLlib 1660 lightgbm 1620 H20 1069 Fastai 848 Caffe 816 catboost 795 Prophet 603 CNTK 541 Mxnet 471 mlr 415

Interesującą zależnością jest tutaj liczba używanych bibliotek uczenia maszynowego używanych w Pythonie oraz R. Caret (najpopularniejsza biblioteka machine learning w R) jest używana ponad 5 razy mniej niż Scikit-Learn (biblioteka z Pythona), podczas gdy osób deklarujących znajomość Pythona było 2 razy więcej niż deklarujących znajomość R. Może to oznaczać, że machine learning jest rzadziej przeprowadzany w R.

In [101]:
avg_cols_salary(38, 55)

Machine learning frameworks Average salary CNTK 62916.0 Caffe 59265.0 Caret 64206.0 Fastai 62506.0 H20 74747.0 Keras 54099.0 Mxnet 70987.0 Prophet 70502.0 PyTorch 53950.0 Scikit-Learn 52944.0 Spark MLlib 70155.0 TensorFlow 53212.0 Xgboost 60515.0 catboost 53314.0 lightgbm 58350.0 mlr 55880.0 randomForest 56950.0

Najwyższe zarobki gwarantują narzędzia H20, Spark MLlib oraz Prophet. 

Rozsądnym podejściem wydaje się więc nauka również mniej znanych rozwiązań niż Scikit-Learn, czy TensorFlow, gdyż specjaliści w mniej znanych programach mogą liczyć na wyższe wynagrodzenie.

### Najczęściej używane narzędzie machine learning

In [104]:
val_counts(57)

Most often used maching learning framework count null 5460 Scikit-Learn 4498 TensorFlow 1355 Keras 1225 randomForest 461 Xgboost 446 Caret 443 PyTorch 437 Spark MLlib 161 lightgbm 128 H20 115 Fastai 102 Other 85 mlr 47 Caffe 45 Mxnet 26 Prophet 24 catboost 21 CNTK 9

In [105]:
avg_salary(57)

Most often used maching learning framework Average salary null 39730.0 CNTK 96389.0 Caffe 40056.0 Caret 59334.0 Fastai 70490.0 H20 86022.0 Keras 49984.0 Mxnet 73462.0 Other 76971.0 Prophet 73229.0 PyTorch 44394.0 Scikit-Learn 52512.0 Spark MLlib 73913.0 TensorFlow 49220.0 Xgboost 73845.0 catboost 33929.0 lightgbm 55527.0 mlr 47660.0 randomForest 56345.0

Wykres sugeruje, że zdecydowanie najwyższe zarobki można uzyskać, jeśli przewodzącym narzędziem do machine learning jest CNTK. Zauważmy jednak, że jedynie dla 9 obserwacji był on głównym narzędziem machine learning, więc nie można uznać tego wyniku za wiarygodny.

Najlepiej opłacanamy narzędziem machine learning wydaje się być H20 z wyraźną przewagą (ponad 10 tysięcy dolarów) nad Prophet, Spark MLlib, XGBoost i Mxnet. Ciekawe może wydawać się drugie miejsce pod względem zarobków u osób korzystających z innych narzędzi machine learning. Potwierdza więc to naszą tezę, że warto znać niszowe rozwiązania machine learning.

### Narzędzia do wizualizacji danych

In [108]:
chart = chart + 1
val_cols_counts(58, 69)

Visualization tools count Matplotlib 9732 ggplot2 5873 Seaborn 5840 Plotly 4315 Shiny 2144 D3 1789 Bokeh 1528 Leaflet 837 Lattice 676 Geoplotlib 540 Altair 171

In [109]:
avg_cols_salary(58, 69)

Visualization tools Average salary Altair 72588.0 Bokeh 66729.0 D3 71330.0 Geoplotlib 46718.0 Lattice 58266.0 Leaflet 62485.0 Matplotlib 49629.0 Plotly 56050.0 Seaborn 53284.0 Shiny 68618.0 ggplot2 57107.0

Według ankietowanych najwyższe zarobki wydają się przynosić 4 narzędzia: Altair, D3, Shiny oraz Bokeh.

Altair jest dość rzadko używaną biblioteką Pythona do wizualizacji danych, więc specjaliści z jej znajomością są wysoko opłacani. Shiny z kolei jest biblioteką umożliwiającą tworzenie aplikacji graficznych w R, a Bokeh w Pythonie. Używanie ich wydaje się być równoważne, jeśli chodzi o zarobki.

### Najczęściej używane narzędzie do wizualizacji danych

In [112]:
val_counts(71)

Most often used visualization framework count null 5930 Matplotlib 4899 ggplot2 2260 Seaborn 981 Plotly 423 D3 198 Shiny 171 Other 86 Bokeh 63 Leaflet 32 Lattice 27 Altair 10 Geoplotlib 8

In [113]:
avg_salary(71)

Most often used visualization framework Average salary null 40667.0 Altair 57750.0 Bokeh 75040.0 D3 71477.0 Geoplotlib 53125.0 Lattice 56481.0 Leaflet 42891.0 Matplotlib 49761.0 Other 79651.0 Plotly 53032.0 Seaborn 53122.0 Shiny 63611.0 ggplot2 60876.0

O dziwo wydaje się, że największe zarobki otrzymują osoby, których główne narzędzie wykorzystywane w pracy jest inne od podanych 11. Pokazuje to cenienie sobie specjalistów w niszowych narzędziach do wizualizacji danych.

Bokeh oraz D3 wydają się być lepiej opłacane niż Shiny, który z kolei znajduje się na podobnym poziomie co ggplot2. Można podejrzewać, że wyniki te są do siebie podobne, gdyż osoby korzystające z tych narzędzi używają R do wizualizacji.

### Relacyjne bazy danych

Niestety jak się okazuje, nazywanie kolumn według najczęściej występującej wartości poskutkowało utworzeniem się 2 kolumn SAP IQ. Aby pozbyć się tego problemu, stwórzmy z kopii danych 2 takie kolumny, ale z różnymi nazwami, a następnie dołączmy je do naszej głównej tabeli z danymi.

Kolumn o nazwie SAP IQ niestety nie możemy podmienić na 2 różne, gdyż są traktowane jako takie same i zmiany na jednej skutkują także takimi samymi zmianami na innej. Dodane nowe kolumny z kolei pojawią się na końcu zbioru danych. Najprościej więc nie korzystać w tym przypadku z funkcji i złączyć części zbioru ręcznie.

In [117]:
data_SAP = data_copy.select('Q29_Part_17', 'Q30_Part_15')

In [118]:
data = data.withColumn("id", monotonically_increasing_id())
data_SAP = data_SAP.withColumn("id", monotonically_increasing_id())

data.createOrReplaceTempView('data1')
data_SAP.createOrReplaceTempView('data2')

data = spark.sql("select * from data1 join data2 on data1.id = data2.id")

In [119]:
data = data.withColumnRenamed('Q29_Part_17', 'SAP IQ database')
data = data.withColumnRenamed('Q30_Part_15', 'SAP IQ analytics')

data = data.drop('id')

In [120]:
chart = chart + 1
results = []

for question in range(72, 88):
  results.append([data.columns[question], data.select(data.columns[question]).filter(data[data.columns[question]] == data.columns[question]).count()])

for question in range(89, 98):
  results.append([data.columns[question], data.select(data.columns[question]).filter(data[data.columns[question]] == data.columns[question]).count()])
  
results.append(['SAP IQ database', data.select(data['SAP IQ database']).filter(data['SAP IQ database'] == data['SAP IQ database']).count()])
  
display(spark.createDataFrame(results).toDF(chart_names[chart], 'count').sort(desc('count')))  

Relational databases count MySQL 4832 PostgresSQL 3201 SQLite 2945 Microsoft SQL Server 2532 Oracle Database 1922 Microsoft Access 1054 AWS Relational Database Service 781 AWS DynamoDB 602 Azure SQL Database 539 Google Cloud SQL 484 Google Cloud Bigtable 427 Google Cloud Datastore 376 AWS Aurora 279 IBM Cloud Db2 196 Azure Database for MySQL 192 Azure Cosmos DB 173 AWS SimpleDB 153 Azure Database for PostgreSQL 106 Google Fusion Tables 92 Google Cloud Spanner 79 SAP IQ database 57 IBM Cloud Compose 44 Ingres 42 IBM Cloud Compose for MySQL 40 IBM Cloud Compose for PostgreSQL 35 NexusDB 25

In [121]:
results = []

for question in range(72, 88):
  query = data.select(data.columns[question], 'Salary_avg').filter(data[data.columns[question]] == data.columns[question]).groupBy(data.columns[question]).avg('Salary_avg')
  results.append([data.columns[question], query.select(bround('avg(Salary_avg)').alias('Average salary')).sort((data.columns[question])).collect()[0]['Average salary']])
  
for question in range(89, 98):
  query = data.select(data.columns[question], 'Salary_avg').filter(data[data.columns[question]] == data.columns[question]).groupBy(data.columns[question]).avg('Salary_avg')
  results.append([data.columns[question], query.select(bround('avg(Salary_avg)').alias('Average salary')).sort((data.columns[question])).collect()[0]['Average salary']])
  
query = data.select(data['SAP IQ database'], 'Salary_avg').filter(data['SAP IQ database'] == data['SAP IQ database']).groupBy(data['SAP IQ database']).avg('Salary_avg')
results.append(['SAP IQ database', query.select(bround('avg(Salary_avg)').alias('Average salary')).sort((data['SAP IQ database'])).collect()[0]['Average salary']])  
  
display(spark.createDataFrame(results).toDF(chart_names[chart], 'Average salary').sort(desc('Average salary')))

Relational databases Average salary AWS Aurora 93790.0 AWS Relational Database Service 87382.0 Google Cloud Spanner 85854.0 AWS DynamoDB 83679.0 Ingres 83274.0 Google Cloud Bigtable 83267.0 IBM Cloud Db2 78559.0 AWS SimpleDB 77533.0 IBM Cloud Compose for PostgreSQL 77357.0 Azure Cosmos DB 76416.0 Azure Database for PostgreSQL 75047.0 IBM Cloud Compose for MySQL 74250.0 IBM Cloud Compose 73068.0 Azure SQL Database 72324.0 Google Fusion Tables 70326.0 Google Cloud SQL 69323.0 Google Cloud Datastore 69229.0 PostgresSQL 66924.0 SAP IQ database 66096.0 Microsoft SQL Server 65542.0 Microsoft Access 63472.0 Azure Database for MySQL 62396.0 SQLite 62294.0 Oracle Database 61387.0 MySQL 58243.0 NexusDB 54800.0

Jak wynika z wykresu, najwyższe zarobki otrzymują osoby korzystające z baz danych chmurowych. Przewodzą wśród nich usługi AWS, przede wszystkim AWS Aurora.

### Używane narzędzia do analiz i Big Data

Wśród zmiennych dotyczących używanych narzędzi do analiz i Big Data również znajduje się kolumna SAP IQ, więc tak jak w poprzednim podejściu nie korzystamy z gotowych funkcji.

In [125]:
chart = chart + 1
results = []

for question in range(100, 114):
  results.append([data.columns[question], data.select(data.columns[question]).filter(data[data.columns[question]] == data.columns[question]).count()])

for question in range(115, 123):
  results.append([data.columns[question], data.select(data.columns[question]).filter(data[data.columns[question]] == data.columns[question]).count()])
  
results.append(['SAP IQ analytics', data.select(data['SAP IQ analytics']).filter(data['SAP IQ analytics'] == data['SAP IQ analytics']).count()])
  
display(spark.createDataFrame(results).toDF(chart_names[chart], 'count').sort(desc('count')))

Big Data and analytics count Google BigQuery 1008 AWS Redshift 828 Databricks 710 AWS Elastic MapReduce 526 Teradata 492 Microsoft Analysis Services 335 AWS Athena 333 Google Cloud Dataflow 313 AWS Kinesis 287 Azure SQL Data Warehouse 225 Google Cloud Pub/Sub 217 Snowflake 203 Azure HDInsight 197 Google Cloud Dataproc 187 AWS Batch 173 Google Cloud Dataprep 164 Oracle Exadata 164 Azure Stream Analytics 137 IBM Cloud Analytics Engine 99 Oracle Warehouse Builder 97 IBM Cloud Streaming Analytics 56 SAP IQ analytics 46 IBM InfoSphere DataStorage 44

In [126]:
results = []

for question in range(100, 114):
  query = data.select(data.columns[question], 'Salary_avg').filter(data[data.columns[question]] == data.columns[question]).groupBy(data.columns[question]).avg('Salary_avg')
  results.append([data.columns[question], query.select(bround('avg(Salary_avg)').alias('Average salary')).sort((data.columns[question])).collect()[0]['Average salary']])
  
for question in range(115, 123):
  query = data.select(data.columns[question], 'Salary_avg').filter(data[data.columns[question]] == data.columns[question]).groupBy(data.columns[question]).avg('Salary_avg')
  results.append([data.columns[question], query.select(bround('avg(Salary_avg)').alias('Average salary')).sort((data.columns[question])).collect()[0]['Average salary']])
  
query = data.select(data['SAP IQ analytics'], 'Salary_avg').filter(data['SAP IQ analytics'] == data['SAP IQ analytics']).groupBy(data['SAP IQ analytics']).avg('Salary_avg')
results.append(['SAP IQ analytics', query.select(bround('avg(Salary_avg)').alias('Average salary')).sort((data['SAP IQ analytics'])).collect()[0]['Average salary']])  
  
display(spark.createDataFrame(results).toDF(chart_names[chart], 'Average salary').sort(desc('Average salary')))

Big Data and analytics Average salary Snowflake 95493.0 AWS Redshift 94438.0 AWS Athena 93341.0 AWS Kinesis 91394.0 IBM InfoSphere DataStorage 89205.0 AWS Elastic MapReduce 87928.0 Teradata 85950.0 IBM Cloud Analytics Engine 81591.0 Databricks 81384.0 Azure HDInsight 75711.0 AWS Batch 74581.0 Google BigQuery 74539.0 Google Cloud Dataflow 71078.0 Google Cloud Dataproc 70829.0 Oracle Exadata 70381.0 Azure SQL Data Warehouse 69589.0 Azure Stream Analytics 68668.0 Google Cloud Pub/Sub 68364.0 IBM Cloud Streaming Analytics 67946.0 SAP IQ analytics 64837.0 Google Cloud Dataprep 63552.0 Microsoft Analysis Services 62746.0 Oracle Warehouse Builder 53557.0

Najlepiej opłacanymi narzędziami do analizy dużych zbiorów danych okazały się być Snowflake, AWS Redshift, AWS Athena, AWS Kinesis, AWS ELastic MapReduce oraz IBM InfoSphere DataStorage. Prym w najlepiej opłacanych narzędziach wiodą więc rozwiązania AWS. Ciekawi również pierwsza pozycja narzędzia Snowflake (i nie wynika to z małej ilości obserwacji).

Databricks występuje w połowie zestawienia, ze stratą około 10 tysięcy dolarów do rozwiązań AWS.

###Używane typy danych

In [129]:
chart = chart + 1
val_cols_counts(125, 137)

Data types count Numerical Data 8177 Text Data 6432 Categorical Data 6146 Time Series Data 5957 Tabular Data 5624 Image Data 3735 Sensor Data 1884 Geospatial Data 1771 Video Data 973 Genetic Data 782 Audio Data 666 Other Data 297

In [130]:
avg_cols_salary(125, 137)

Data types Average salary Audio Data 50447 Categorical Data 56914 Genetic Data 50604 Geospatial Data 66522 Image Data 45935 Numerical Data 54222 Other Data 55429 Sensor Data 59816 Tabular Data 58600 Text Data 54049 Time Series Data 60033 Video Data 48864

Najwyższe zarobki pojawiały się u ankieterów, którzy mają w pracy do czynienia z danymi przestrzennymi, szzeregami czasowymi oraz danymi sensorycznymi.

## Podsumowanie

Przeprowadzona została transformacja danych związanych z branżą Data Science, a później ich agregacja ze względu na średnie zarobki respondendentów.

Dużą część projektu zajęła transformormacja danych do postaci umożliwiającej sprawną agregację wielu kolumn danych. Faza ta była konieczna do uzyskania przejrzystych wyników agregacji.

Najważniejsze zmienne dotyczące charakterystyk zadań wykonywanych w Data Science zostały również przedstawione na wykresach w odniesieniu do przeciętnych zarobków osób posiadających daną charakterystykę. Aby upewnić się, że dane wyniki są wiarygodne, dokonano także zliczenia rekordów unikatowych danych, aby nie generalizować wyników na zbyt małych próbach danych.